service/storage_service.cc

```diff
@@ -742,12 +751,41 @@ future<> storage_service::topology_change_coordinator_fiber(raft::server& raft,
                         auto tmptr = get_token_metadata_ptr();
                         auto bootstrap_tokens = boot_strapper::get_random_bootstrap_tokens(tmptr, num_tokens, dht::check_token_endpoint::yes);
 
-                        // Write choosen tokens through raft.
+                        auto get_sharding_info = [] (dht::token) -> std::pair<size_t, uint8_t> {
+                            // FIXME: store shard count and ignore_msb for each node in topology, use them here.
+                            return {1, 12};
+                        };
+                        auto [gen_uuid, gen_desc] = cdc::make_new_generation_data(
+                            bootstrap_tokens, get_sharding_info, tmptr);
+                        auto gen_table_schema = _db.local().find_schema(
+                            db::system_keyspace::NAME, db::system_keyspace::CDC_GENERATIONS_V3);
+
+                        // FIXME: the CDC generation data can be large and not fit in a single command
+                        // (for large clusters, it will introduce reactor stalls and go over commitlog entry
+                        // size limit). We need to split it into multiple mutations by smartly picking
+                        // a `mutation_size_threshold` and sending each mutation as a separate group 0 command.
+                        // We also don't want to serialize the commands - there may be many of them,
+                        // and we don't want to wait for a network round-trip to a quorum between each command.
+                        // So we need to introduce a mechanism for group 0 to send a sequence of commands
+                        // that can be committed concurrently. Also we need to be careful with memory consumption
+                        // with many large mutations.
+                        // See `system_distributed_keyspace::insert_cdc_generation` for inspiration how it
+                        // was done when the mutations were stored in a regular distributed table.
+                        const size_t mutation_size_threshold = 2'000'000;
+                        auto gen_mutations = co_await cdc::get_cdc_generation_mutations(
+                            gen_table_schema, gen_uuid, gen_desc, mutation_size_threshold, node.guard.write_timestamp());
+                        std::vector<canonical_mutation> updates{gen_mutations.begin(), gen_mutations.end()};
```

Comment:
Which logic though? The details of how CDC generations are created and serialized into mutations are already abstracted by these `cdc` namespace functions.

The logic here is to put the mutations inside group 0 updates (together with the `builder.build()` mutation below), which I think makes sense to be part of the coordinator code.

Maybe you have something like this in mind:
```
std::vector<canonical_mutation> updates;
auto gen_uuid = co_await cdc::make_new_generation_and_put_as_mutations(updates, ...);
builder.(...).set("new_cdc_generation_data_uuid", gen_uuid);
updates.push_back(builder.build());
```

but I don't want to do that because of the FIXME above - the final version of the code (before leaving experimental at least) will have to send multiple group 0 commands, so the `make_new_generation_and_put_as_mutations` abstraction won't fit.

Reply:
Ok, I'm not going to insist now.